pubtools/_ami/tasks/push.py

```diff
@@ -210,32 +215,12 @@ def update_rhsm_metadata(self, image, push_item):
     def _push_to_region(self, region_data):
         # sends the push_items for each region to be uploaded in
         # a single thread
-        retry_wait = self.args.retry_wait or 30
-        max_retries = self.args.max_retries or 4
-
         for dest_data in region_data:
```

Comment:
I've just noticed that the movement of retry logic outside of this method means that the overall retry logic will have changed quite a bit.

Previously it was inside of the loop here.  I'm not sure what region_data normally looks like, but let's say it contained [A, B, C] and B failed on the first try.  Then the sequence of events would have been:

- upload A
- upload B (attempt 1 - failed)
- upload B (attempt 2 - succeeded)
- upload C

But now with the retry happening at a higher level, it's going to retry the entire loop, isn't it?  Like this?

- upload A
- upload B (attempt 1 - failed)
- upload A (again)
- upload B (attempt 2 - succeeded)
- upload C

It's hard for me to understand what the impact of this is going to be as I don't know what's usually in `region_data`.  The variable names are not helping at all (`region_data` seems to be a list here, but in the calling method there is another variable also named `region_data` which instead seems to be a list of list?)

Anyway, have you considered that?  Would it be harmful that the retries are now going to cover the entire `for dest_data in region_data:` loop rather than covering individual items within the region_data list?

Reply:
Thanks for pointing that out. I don't have enough data to verify that this will not have a big performance impact, so I added a check to verify that the image is not already pushed before pushing it.